crates/fluvio/src/sync/mod.rs

```diff
@@ -108,16 +108,26 @@ mod context {
             use tokio::select;
             use fluvio_future::timer::sleep;
 
+            // We can short circuit here if already present
+            let found = search(self.store().read().await);
+            if found.is_some() {
+                return Ok(found);
+            }
+
             let mut timer = sleep(Duration::from_millis(*MAX_WAIT_TIME));
             let mut listener = self.store.change_listener();
 
             // wait for either changes from store or timeout
             select! {
 
                 _ = listener.listen() => {
+                    // since the listeners change count is always 0 and listener.listen() only returns when change count > 0,
+                    // calling for sync_changes() will always trigger an allocation.
+                    // Since we don't do anything with the changes and sync_changes() doesn't add new changes to the cache, it's wasteful to call
+
                     // this should be full sync
-                    let changes = listener.sync_changes().await;
```

Comment:
`sync_changes` modifies the internal state here: https://github.com/infinyon/fluvio/blob/92cece6daf0c878cabf0210b3f78b00e739006e2/crates/fluvio-stream-model/src/store/dual_store.rs#L539. Are you sure it's completely safe to delete this call here?

Reply:
@sehz Could you pls take a look

Reply:
From what I can see, `self.set_last_change(...)` only modifies the `last_change` field on the calling `ChangeListener`. In this case, the `ChangeListener` is only is scope for the `lookup_and_wait` function call, and then gets dropped before the field is ever read.

> `sync_changes` modifies the internal state here:
> 
> https://github.com/infinyon/fluvio/blob/92cece6daf0c878cabf0210b3f78b00e739006e2/crates/fluvio-stream-model/src/store/dual_store.rs#L539
> . Are you sure it's completely safe to delete this call here?

Reply:
I think it would be necessary to call `sync_changes()`  if `listen()` was called in a loop wanting to be woken up with new changes, but here is only called once.

Reply:
It's been a while since I looked at this part code.
Meanwhile, can you add tests to verify these changes are safe?

Reply:
I added a refactor to make it explicit that calling `lookup_by_key` doesn't do anything that could mutate the LocalStore's state. It no longer requires a mutable reference for a `ChangeListener` nor access to LocalStore's internal state. I also tightened up the interfaces a bit for the LocalStore and EventPublisher. 

As far as testing, what I would normally recommend is to have a set of unit tests for all of the public methods on interfaces for the synchronization structs and then protect access to the internals. I didn't remove or touch any preexisting tests for `lookup_by_key` or for the synced state, so if those already exist then the change is safe. If they don't exist, tightening up the interfaces and writing comprehensive unit tests for them would be a bit of a refactor, is that a the direction you want me to take?

Reply:
No, just enough unit test to cover changes you proposed here.

Reply:
But this seems good changes.   Good work

Reply:
Great.  Nice testing